packages/components/src/navigation/index.js

```diff
@@ -1,43 +1,65 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
 /**
  * WordPress dependencies
  */
 import { useEffect, useState } from '@wordpress/element';
+import { usePrevious } from '@wordpress/compose';
 
 /**
  * Internal dependencies
  */
+import Animate from '../animate';
 import { Root } from './styles/navigation-styles';
 import Button from '../button';
 
 const Navigation = ( { activeItemId, children, data, rootTitle } ) => {
-	const [ activeLevel, setActiveLevel ] = useState( 'root' );
+	const [ activeLevelId, setActiveLevelId ] = useState( 'root' );
+
+	const appendItemData = ( item ) => {
+		return {
+			...item,
+			children: [],
+			parent: item.id === 'root' ? null : item.parent || 'root',
+			isActive: item.id === activeItemId,
+			setActiveLevelId,
+		};
+	};
 
-	const mapItemData = ( items ) => {
-		return items.map( ( item ) => {
-			const itemChildren = data.filter( ( i ) => i.parent === item.id );
-			return {
-				...item,
-				children: itemChildren,
-				parent: item.parent || 'root',
-				isActive: item.id === activeItemId,
-				hasChildren: itemChildren.length > 0,
-				setActiveLevel,
-			};
+	const mapItems = ( itemData ) => {
+		const items = new Map(
+			[
+				{ id: 'root', parent: null, title: rootTitle },
+				...itemData,
+			].map( ( item ) => [ item.id, appendItemData( item ) ] )
+		);
+
+		items.forEach( ( item ) => {
+			const parentItem = items.get( item.parent );
+			if ( parentItem ) {
+				parentItem.children.push( item );
+				parentItem.hasChildren = true;
+			}
 		} );
+
+		return items;
 	};
-	const items = [ { id: 'root', title: rootTitle }, ...mapItemData( data ) ];
 
-	const activeItem = items.find( ( item ) => item.id === activeItemId );
-	const level = items.find( ( item ) => item.id === activeLevel );
-	const levelItems = items.filter( ( item ) => item.parent === level.id );
-	const parentLevel =
-		level.id === 'root'
-			? null
-			: items.find( ( item ) => item.id === level.parent );
+	const items = mapItems( data );
```

Comment:
This calculation happens on every render. How about adding [useMemo](https://github.com/WordPress/gutenberg/tree/master/packages/element#useMemo) here?

Reply:
Great idea!  Added in https://github.com/WordPress/gutenberg/pull/24771/commits/e6872669b92073cf5d6394040645deaed28aecbe.

I think there are some further optimizations that we could tackle here so a full update doesn't occur whenever the `activeItemId` is changed, but may need to handle in a follow-up.